opamp/observiq/observiq_downloadable_file_manager.go

```diff
@@ -171,3 +175,146 @@ func (m DownloadableFileManager) CleanupArtifacts() {
 		m.logger.Error("Failed to remove temporary directory", zap.Error(err))
 	}
 }
+
+// unarchive will unpack the package at archivePath(.tar.gz or .zip) into the directory found at extractPath
+func unarchive(archivePath, extractPath string) error {
+	if strings.HasSuffix(archivePath, ".tar.gz") {
+		// Handle tar.gz files
+		if err := extractTarGz(archivePath, extractPath); err != nil {
+			return fmt.Errorf("extract tar.gz: %w", err)
+		}
+	} else if strings.HasSuffix(archivePath, ".zip") {
+		// Handle zip files
+		if err := extractZip(archivePath, extractPath); err != nil {
+			return fmt.Errorf("extract .zip: %w", err)
+		}
+	} else {
+		return fmt.Errorf("unsupported file type: %s", archivePath)
+	}
+
+	return nil
+}
+
+// extractTarGz will extract the .tar package at archivePath into the dir at extractPath
+func extractTarGz(archivePath, extractPath string) error {
+	if err := os.MkdirAll(extractPath, 0750); err != nil {
+		return fmt.Errorf("mkdir extract path: %w", err)
+	}
+
+	archivePathClean := filepath.Clean(archivePath)
+	file, err := os.Open(archivePathClean)
+	if err != nil {
+		return fmt.Errorf("open archive package: %w", err)
+	}
+	defer file.Close()
+
+	gzipReader, err := gzip.NewReader(file)
+	if err != nil {
+		return fmt.Errorf("new gzip reader: %w", err)
+	}
+	defer gzipReader.Close()
+
+	tarReader := tar.NewReader(gzipReader)
+
+	for {
+		header, err := tarReader.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return fmt.Errorf("read next tarball header: %w", err)
+		}
+
+		outputPath, err := sanitizeArchivePath(extractPath, header.Name)
+		if err != nil {
+			return fmt.Errorf("sanitize archive path: %w", err)
+		}
+
+		switch header.Typeflag {
+		case tar.TypeDir:
+			if err := os.MkdirAll(outputPath, os.FileMode(header.Mode)); err != nil {
+				return fmt.Errorf("mkdir: %w", err)
+			}
+
+		case tar.TypeReg:
+			outputPathClean := filepath.Clean(outputPath)
+			outFile, err := os.Create(outputPathClean)
+			if err != nil {
+				return fmt.Errorf("create file: %w", err)
+			}
+			defer outFile.Close()
+
+			_, err = io.CopyN(outFile, tarReader, maxArchiveObjectByteSize)
+			if err != nil && err != io.EOF {
+				return fmt.Errorf("write to file: %w", err)
+			}
```

Comment:
As I understand it, CopyN returns an EOF error if the bytes to write (maxArchiveObjectBytesSize in this case) is larger than the source (tarReader here). I think it's safe to ignore this EOF error since we can't know for sure the exact sizes of these source files.